docs/styleguide-cpp.md

```diff
@@ -1,320 +1,660 @@
-	```c++
-	//Correct
-	bool isEven(int num) {
-		return num % 2 == 0;
-	}
-
-	//Incorrect
-	bool even(int num) {
-		return num % 2 == 0;
-	}
-	```
-- When a function parameter is being passed by const referance, put one space between the type and `const&`. If a parameter is passed by mutable reference do not put a space between the type and the `&`
-	```c++
-	//Correct for a const reference
-	bool hasLargeNumber(std::vector<int> const& numbers) {
-		//some cool stuff
-		return true;
-	}
-
-	//Incorrect for a const reference
-	bool hasLargeNumber(std::vector<int>const& numbers) {
-		//some cool stuff
-		return true;
-	}
-
-	//Correct for a mutable reference
-	bool hasLargeNumber(std::vector<int>& numbers) {
-		//some cool stuff
-		return true;
-	}
-
-	//Incorrect for a mutable reference
-	bool hasLargeNumber(std::vector<int> & numbers) {
-		//some cool stuff
-		return true;
-	}
-	```
--  The include order should be: C++ standard headers, followed by Godot headers, then OpenVic2 headers
-	- C++ standard headers and Godot headers should be enclosed by angle brackets `<>`
-	- OpenVic2 headers should be enclosed by double quotes `""`
-	```c++
-	//Correct
-	#include <vector>
-	#include <string>
-	#include <godot_cpp/classes/object.hpp>
-	#include <godot_cpp/core/class_db.hpp>
-	#include "openvic2/src/SampleHeader.hpp"
-	#include "openvic2/src/SampleHeader2.hpp"
-
-	//Incorrect
-	#include "openvic2/src/SampleHeader.hpp"
-	#include <godot_cpp/classes/object.hpp>
-	#include "openvic2/src/SampleHeader2.hpp"
-	#include <godot_cpp/core/class_db.hpp>
-	#include <vector>
-	#include <string>
-	```
+2.  C standard headers
+
+3.  C++ standard headers
+
+4.  Godot headers
+
+5.  OpenVic2 dependency headers
+
+6.  OpenVic2 headers
+
+Do’s and Don’ts
+
+1.  Related implementation headers should be enclosed by double quotes
+    `""`
+
+    - Related implementation headers should be relative to its directory
+      and should always reside in the same directory.
+
+    - Related implementation headers should be separated from other
+      headers.
+
+2.  C standard headers, C++ standard headers, and Godot headers should
+    be enclosed by angle brackets `<>`
+
+3.  OpenVic2 dependency headers and OpenVic2 should be enclosed by
+    double quotes `""`
+
+```c++
+//Correct
+#include "Implementation.hpp"
+
+#include <stdio>
+#include <vector>
+#include <string>
+#include <godot_cpp/classes/object.hpp>
+#include <godot_cpp/core/class_db.hpp>
+#include ""
+#include "openvic2/src/SampleHeader.hpp"
+#include "openvic2/src/SampleHeader2.hpp"
+
+//Incorrect
+#include "openvic2/src/SampleHeader.hpp"
+#include <godot_cpp/classes/object.hpp>
+#include "openvic2/src/SampleHeader2.hpp"
+#include <stdio>
+#include <godot_cpp/core/class_db.hpp>
+#include <vector>
+#include <string>
+#include "Implementation.hpp"
+```
+
+#### 1.4.23. Prefer Godot’s Variant types over Godot’s template types and the Standard Library
```

Comment:
Also what about code that doesnt use godot?

Reply:
Reliance on Godot is inevitable if you're gonna use a game engine.

Reply:
Using some of these types, and hence not using the standard library types that do the same things, is going to make it extremely hard to use anything else in the C++ ecosystem

Reply:
> Reliance on a game engine is inevitable if you're gonna use a game engine.

There is a difference between relying on Godot for rendering/sound/UI and making our core simulation dependant on the internals of Godot's engine

For example, if somebody wanted to write an OpenGL UI/frontend for OpenVic2, it would make no sense for the C++ simulation to be reliant on non-standard Godot types

Reply:
It says prefer for a reason.